Source/SpatialGDK/Public/Utils/SchemaOption.h

```diff
@@ -0,0 +1,108 @@
+#pragma once
+
+#include <improbable/c_worker.h>
+#include "UniquePtr.h"
+
+template <typename T>
+class TSchemaOption
+{
+public:
+	TSchemaOption() = default;
+	~TSchemaOption() = default;
+
+	TSchemaOption(const T& InValue)
+	{
+		Value = MakeUnique<T>(InValue);
+	}
+
+	TSchemaOption(T&& InValue)
+	{
+		Value = MakeUnique<T>(MoveTemp(InValue));
+	}
+
+	TSchemaOption(const TSchemaOption& InValue)
+	{
+		*this = InValue;
+	}
+
+	TSchemaOption(TSchemaOption&&) = default;
+
+	TSchemaOption& operator=(const TSchemaOption& InValue)
+	{
+		if (this != &InValue)
+		{
+			if (InValue)
+			{
+				Value = MakeUnique<T>(*InValue);
+			}
+			else
+			{
+				Value.Reset();
+			}
+		}
+
+		return *this;
+	}
+
+	TSchemaOption& operator=(TSchemaOption&&) = default;
+
+	FORCEINLINE bool IsSet() const
+	{
+		return Value.IsValid();
+	}
+
+	FORCEINLINE explicit operator bool() const
+	{
+		return IsSet();
+	}
+
+	const T& GetValue() const
+	{
+		checkf(IsSet(), TEXT("It is an error to call GetValue() on an unset TSchemaOptional. Please check IsSet()."));
+		return *Value;
+	}
+
+	T& GetValue()
+	{
+		checkf(IsSet(), TEXT("It is an error to call GetValue() on an unset TSchemaOptional. Please check IsSet()."));
+		return *Value;
+	}
+
+	bool operator==(const TSchemaOption& InValue) const
+	{
+		if (IsSet() != InValue.IsSet())
+		{
+			return false;
+		}
+
+		if (!IsSet())
+		{
+			return true;
+		}
+
+		return GetValue() == InValue.GetValue();
+	}
+
+	bool operator!=(const TSchemaOption& InValue) const
+	{
+		return !operator==(InValue);
+	}
+
+	T& operator*() const
+	{
+		return *Value;
+	}
+
+	const T* operator->() const
+	{
+		return Value.Get();
+	}
+
+	T* operator->()
+	{
+		return Value.Get();
+	}
+
+private:
+	TUniquePtr<T> Value;
+};
```

Comment:
Should we also implement a `GetTypeHash` on this and make the `GetTypeHash` of `UnrealObjectRef` call it?

Reply:
I don't think so. The implementation of UnrealObjectRef's `GetTypeHash` wouldn't change much with the addition of that function, which would just call `T`'s type hash.

Reply:
From
```
Result = (Result * 977u) + (ObjectRef.Path ? 1327u * (GetTypeHash(*ObjectRef.Path) + 977u) : 977u);
Result = (Result * 977u) + (ObjectRef.Outer ? 1327u * (GetTypeHash(*ObjectRef.Outer) + 977u) : 977u);```
to
```
Result = (Result * 977u) + GetTypeHash(ObjectRef.Path);
Result = (Result * 977u) + GetTypeHash(ObjectRef.Outer);```

Reply:
Ok I can add it.